pkg/policy/api/ingress.go

```diff
@@ -65,6 +65,15 @@ type IngressCommonRule struct {
 	// +kubebuilder:validation:Optional
 	FromCIDRSet CIDRRuleSlice `json:"fromCIDRSet,omitempty"`
 
+	// FromCIDRGroupRef is a list of CIDRs which the endpoint subject to the
+	// rule is allowed to receive connections from in addition to FromEndpoints,
+	// along with a list of subnets contained within their corresponding IP block
+	// from which traffic should not be allowed.
+	// This will match on the source IP address of incoming connections.
+	//
+	// +kubebuilder:validation:Optional
+	FromCIDRGroupRef CIDRGroupRefSlice `json:"fromCIDRGroupRefs,omitempty"`
```

Comment:
Embedding `cidrGroupRef` under `fromCIDRSet` has a couple of ambiguities, too:

- In the CIDRRule, the field CIDR is [required](https://github.com/cilium/cilium/blob/master/pkg/policy/api/cidr.go#L41). Changing that to optional to enable the cidr group reference embedding will introduce the possibility to have empty CIDRRules.
- The Ingress rule sanitization [does not allow](https://github.com/cilium/cilium/blob/master/pkg/policy/api/rule_validation.go#L103) combining `FromCIDR` and `FromCIDRSet`. This is unfortunate, since after the translation of the cidr group reference in `FromCIDRSet` we add a list of CIDRs right in `FromCIDR`.

Reply:
Hmm I see. Technically it would be possible to support this with `oneOf` semantics, but it looks like the kube-builder code generator we use doesn't support that statement.

One alternative would be to move this code later in the process to a post-parsing stage so that the effects of this translation do not need to be subject to the same checks as are being enforced at the API layer here.

Incidentally, how does the 'there must be at least one CIDR' check pass today with an empty CIDR group?

Another option may be to not use this suggestion, but instead just add validation code (+ API comments) to prevent the combination that I suggested above. The concern I had is around the combination of specific statements, but we could define that if there is a CIDRGroupRef then the user cannot also specify CIDRs/CIDRSets. Then it's "just" up to the internal logic to skip that API check iff the CIDRs are generated.

Reply:
After an out-of-band discussion and some experimentation, we decided to embed the field `fromCidrGroup` in the `fromCIDRSet` Ingress rule as suggested.
Regarding the translation, for each referenced cidr group, the CNP/CCNP/CCG handler generates additional `fromCIDRSet` objects containing the CIDRs belonging to the groups.